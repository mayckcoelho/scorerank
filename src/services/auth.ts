import api from "./api";
import ISignInRequest from "../interfaces/ISignInRequest";
import ISignInResponse from "../interfaces/ISignInResponse";
import IRegisterRequest from "../interfaces/IRegisterRequest"

export async function signIn(data: ISignInRequest): Promise<ISignInResponse> {
    return await api.post("/login", data)
}

export async function register(data: IRegisterRequest): Promise<ISignInResponse> {
    return await api.post("/register", data)
}