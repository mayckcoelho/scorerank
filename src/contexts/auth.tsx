import React, { useState, useEffect, useContext, createContext } from 'react'
import jwtDecode from "jwt-decode"
import Toast from 'react-native-simple-toast';
import AsyncStorage from '@react-native-community/async-storage'
import * as auth from '../services/auth'
import api from '../services/api'

import IUser from "../interfaces/IUser";
import IAuthContextData from "../interfaces/IAuthContextData";
import ISignInRequest from "../interfaces/ISignInRequest";
import IRegisterRequest from "../interfaces/IRegisterRequest";

const AuthContext = createContext<IAuthContextData>({} as IAuthContextData);

export const AuthProvider: React.FC = ({ children }) => {
    const [user, setUser] = useState<null | IUser>(null)
    const [loading, setLoading] = useState(false);

    useEffect(() => {
        async function loadStorageData() {
            const [[, storagedUser], [, storagedToken]] = await AsyncStorage.multiGet(['@RNAuth:user', '@RNAuth:token']);

            if (storagedUser && storagedToken) {
                api.defaults.headers.Authorization = `Bearer ${storagedToken}`;

                setUser(JSON.parse(storagedUser));
                setLoading(false);
            }
        }

        loadStorageData();
    }, [])

    const signIn = async (data: ISignInRequest) => {
        const response = await auth.signIn(data).catch(err => {
            Toast.show(err.response.data)
        })

        if (response)
            handleResponse(response.data)
    }

    const register = async (data: IRegisterRequest) => {
        const response = await auth.register(data).catch(err => {
            Toast.show(err.response.data)
        })

        if (response)
            handleResponse(response.data)
    }

    const handleResponse = async (data: any) => {
        const token = data.accessToken;
        const decoded = jwtDecode(token) as {
            sub: string;
            email: string;
        };

        api.defaults.headers.Authorization = `Bearer ${token}`;

        const response = await api.get(`/users/${decoded.sub}`)

        const user = response.data as IUser;
        await AsyncStorage.setItem('@RNAuth:user', JSON.stringify(user))

        setUser(user)
        await AsyncStorage.setItem('@RNAuth:token', token)
    }

    const signOut = () => {
        AsyncStorage.clear().then(() => {
            setUser(null);
        });
    }

    const updateScore = async () => {
        if (user) {
            const newUser = { ...user, score: user.score + 1 }

            await api.patch(`/users/${user?.id}`, {
                score: newUser.score
            })

            await AsyncStorage.setItem('@RNAuth:user', JSON.stringify(newUser))

            setUser(newUser)
        }
    }

    return (
        <AuthContext.Provider value={{ signed: !!user, loading, user, register, signIn, signOut, updateScore }}>
            {children}
        </AuthContext.Provider>
    )
}

export function useAuth() {
    const context = useContext(AuthContext);

    return context;
}