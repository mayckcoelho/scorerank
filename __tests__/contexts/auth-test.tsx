import React from 'react';
import { View, Text } from "react-native"
import { AuthProvider } from '../../src/contexts/auth';
import renderer from 'react-test-renderer';

describe('Auth should render correctly', () => {
    it('match snapshot', () => {

        const auth = renderer
            .create(
                <AuthProvider>
                    <View>
                        <Text>Foo</Text>
                    </View>
                </AuthProvider>)
            .toJSON();
        expect(auth).toMatchSnapshot();
    });
})