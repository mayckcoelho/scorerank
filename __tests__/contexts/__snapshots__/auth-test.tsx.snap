// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth should render correctly match snapshot 1`] = `
<View>
  <Text>
    Foo
  </Text>
</View>
`;
